src/achievements.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+	Copyright (C) 2003 - 2022
+	by David White <dave@whitevine.net>
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#include "achievements.hpp"
+
+#include "filesystem.hpp"
+#include "game_config.hpp"
+#include "log.hpp"
+#include "preferences/general.hpp"
+#include "serialization/parser.hpp"
+#include "serialization/preprocessor.hpp"
+
+static lg::log_domain log_config("config");
+#define ERR_CONFIG LOG_STREAM(err, log_config)
+
+// TODO: horizontal scrollbar is present despite `horizontal_scrollbar_mode = "never"`
+// TODO: how to make the initial height more than a single achievement? hard code the height? add more achievements?
+// TODO: make checkbox to indicate which achievements have been achieved disabled/untoggleable
+// TODO: GUIWindowDefinitionWML in achievements_dialog.hpp
+// TODO: some way to display in-game when something is achieved?
```

Comment:
I want to say a modeless dialog, but as far as I know those still don't work over the map…

Reply:
My thought is that ideally it'd be something that doesn't interrupt what the player is currently doing by forcing them to click to dismiss it, but I don't think (?) Wesnoth current has something like that.

Reply:
Right, which is why I suggested a modeless dialog, which could pop up in the corner, then disappear a few seconds later. I think that would be the ideal form of the notification. However, as far as I know, event handling on the map is still incompatible with GUI2 event handling – the mere existence of a modeless window onscreen will prevent the map from receiving any events at all. So, unless that's solved, we'll have to use a modal dialog, and the object popup seems like the best choice for that.

Reply:
Random thought: could floating text potentially be used for this? There'd need to be a way to add a "background" color behind the text to make it more visible and readable, but that wouldn't block input to the rest of the map while it's displayed.

Reply:
Yes, I think that would be doable. You could make the text relatively large and have it appear in the corner of the screen. The only downside of this approach is that floating text is just text. I don't think there's any way to throw in the icon.